project_euler/problem_234/sol1.py

```diff
@@ -17,40 +17,98 @@
 What is the sum of all semidivisible numbers not exceeding 999966663333 ?
 """
 
+import math
 
-def fib(a, b, n):
-
-    if n == 1:
-        return a
-    elif n == 2:
-        return b
-    elif n == 3:
-        return str(a) + str(b)
-
-    temp = 0
-    for x in range(2, n):
-        c = str(a) + str(b)
-        temp = b
-        b = c
-        a = temp
-    return c
-
-
-def solution(n):
-    """Returns the sum of all semidivisible numbers not exceeding n."""
-    semidivisible = []
-    for x in range(n):
-        l = [i for i in input().split()]  # noqa: E741
-        c2 = 1
-        while 1:
-            if len(fib(l[0], l[1], c2)) < int(l[2]):
-                c2 += 1
-            else:
+
+def sieve(n: int) -> list:
+    """
+    Sieve of Erotosthenes
+    Function to return all the prime numbers up to a certain number
+    https://en.wikipedia.org/wiki/Sieve_of_Eratosthenes
+    >>> sieve(3)
+    [2]
+    >>> sieve(50)
+    [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47]
+    """
+    is_prime = [True] * n
+    is_prime[0] = False
+    is_prime[1] = False
+    is_prime[2] = True
+
+    for i in range(3, int(n ** 0.5 + 1), 2):
+        index = i * 2
+        while index < n:
+            is_prime[index] = False
+            index = index + i
+
+    primes = [2]
+
+    for i in range(3, n, 2):
+        if is_prime[i]:
+            primes.append(i)
+
+    return primes
+
+
+def solution() -> int:
+    """
+    Computes the solution to the problem
+    >>> solution()
+    1259187438574927161
+    """
+    limit = 999966663333
+    primes_upper_bound = math.floor(math.sqrt(limit)) + 100
+    primes = sieve(primes_upper_bound)
+
+    matches_sum = 0
+    prime_index = 0
+    last_prime = primes[prime_index]
+
+    while (last_prime * last_prime) <= limit:
+        next_prime = primes[prime_index + 1]
+
+        lower_bound = last_prime * last_prime
+        upper_bound = next_prime * next_prime
+
+        # Get numbers divisible by lps(current)
+        current = lower_bound + last_prime
+        while current < upper_bound and current <= limit:
+            matches_sum += current
+            current += last_prime
+
+        # Reset the upper_bound
+        while (upper_bound - next_prime) > limit:
+            upper_bound -= next_prime
+
+        # Add the numbers divisible by ups(current)
+        current = upper_bound - next_prime
+        while current > lower_bound:
+            matches_sum += current
+            current -= next_prime
+
+        # Remove the numbers divisible by both ups and lps
+        current = lower_bound - lower_bound % (last_prime * next_prime)
```

Comment:
```suggestion
        current = 0
```
What happened in here?

Reply:
In theory this was to optimize the amount of items to be checked since the `%` operator has precedence over the `-` operator but this ended up not making much of a difference. I think for clarity sake we should indeed start `current` at `0`.